packages/adblocker/src/filters/cosmetic.ts

```diff
@@ -872,6 +879,25 @@ export default class CosmeticFilter implements IFilter {
     return getBit(this.mask, COSMETICS_MASK.remove);
   }
 
+  public isHas(): boolean {
+    if (!this.isExtended()) {
+      return false;
+    }
+    const extendedPseudoClasses = new Set();
+    const ast = this.getSelectorAST();
+
+    walk(ast, (node) => {
+      if (
+        node.type === 'pseudo-class' &&
+        node.name !== undefined &&
+        EXTENDED_PSEUDO_CLASSES.has(node.name)
+      ) {
+        extendedPseudoClasses.add(node.name);
+      }
+    });
+    return extendedPseudoClasses.size === 1 && extendedPseudoClasses.has('has');
```

Comment:
alternatively we can do this detection on `parse` and store a result on a mask `COSMETICS_MASK.isHas`

Reply:
In principle, the implementation in this function itself could be optimized, though I don't know if it is a critical optimization and how big the abstract syntax trees are in practice.

(If it is a hot path, one idea would be this: instead of building a set and in the end test if it has exactly one expected value, it could scan over the tree and look for this value: if it find the first different value, it can immediately exit; otherwise, it can check in the end, if the expected value was present. So, instead of a Set, a bool would be enough.)